File: scripts/reference-solutions/valid-palindrome.py
import sys

s = sys.stdin.read()
if s.endswith("\n"):
    s = s[:-1]
if s.endswith("\r"):
    s = s[:-1]

left = 0
right = len(s) - 1

while left < right:
    while left < right and not s[left].isalnum():
        left += 1
    while left < right and not s[right].isalnum():
        right -= 1
    if s[left].lower() != s[right].lower():
        print("false")
        break
    left += 1
    right -= 1
else:
    print("true")

File: scripts/reference-solutions/valid-palindrome.cpp
#include <cctype>
#include <iostream>
#include <string>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  string s;
  getline(cin, s, '\0');
  if (!s.empty() && s.back() == '\n') {
    s.pop_back();
  }
  if (!s.empty() && s.back() == '\r') {
    s.pop_back();
  }

  int left = 0;
  int right = static_cast<int>(s.size()) - 1;

  while (left < right) {
    while (left < right && !isalnum(static_cast<unsigned char>(s[left]))) {
      ++left;
    }
    while (left < right && !isalnum(static_cast<unsigned char>(s[right]))) {
      --right;
    }
    char a = static_cast<char>(tolower(static_cast<unsigned char>(s[left])));
    char b = static_cast<char>(tolower(static_cast<unsigned char>(s[right])));
    if (a != b) {
      cout << "false\n";
      return 0;
    }
    ++left;
    --right;
  }

  cout << "true\n";
  return 0;
}
